From 7ac0febb81259fb9e0c447e0b90d0f05f409a02b Mon Sep 17 00:00:00 2001
From: David Disseldorp <ddiss@suse.de>
Date: Fri, 28 Jun 2013 11:47:33 +0200
Subject: [PATCH] cifs: fill TRANS2_QUERY_FILE_INFO ByteCount fields
References: bnc#804950
Patch-mainline: v3.11
Git-commit: 7ac0febb81259fb9e0c447e0b90d0f05f409a02b

Currently the trans2 ByteCount field is incorrectly left zero in
TRANS2_QUERY_FILE_INFO info_level=SMB_QUERY_FILE_ALL_INFO and
info_level=SMB_QUERY_FILE_UNIX_BASIC requests. The field should properly
reflect the FID, information_level and padding bytes carried in these
requests.

Leaving this field zero causes such requests to fail against Novell CIFS
servers. Other SMB servers (e.g. Samba) use the parameter count fields
for data length calculations instead, so do not suffer the same fate.

Signed-off-by: David Disseldorp <ddiss@suse.de>
Acked-by: Jeff Layton <jlayton@redhat.com>
Signed-off-by: Steve French <smfrench@gmail.com>
---
 fs/cifs/cifssmb.c | 2 ++
 1 file changed, 2 insertions(+)

Index: linux-3.0-bnc804950_qfileinfo_bytecount_11sp2/fs/cifs/cifssmb.c
===================================================================
--- linux-3.0-bnc804950_qfileinfo_bytecount_11sp2.orig/fs/cifs/cifssmb.c
+++ linux-3.0-bnc804950_qfileinfo_bytecount_11sp2/fs/cifs/cifssmb.c
@@ -3637,6 +3637,7 @@ QFileInfoRetry:
 	pSMB->Pad = 0;
 	pSMB->Fid = netfid;
 	inc_rfc1001_len(pSMB, byte_count);
+	pSMB->t2.ByteCount = cpu_to_le16(byte_count);
 
 	rc = SendReceive(xid, tcon->ses, (struct smb_hdr *) pSMB,
 			 (struct smb_hdr *) pSMBr, &bytes_returned, 0);
@@ -3805,6 +3806,7 @@ UnixQFileInfoRetry:
 	pSMB->Pad = 0;
 	pSMB->Fid = netfid;
 	inc_rfc1001_len(pSMB, byte_count);
+	pSMB->t2.ByteCount = cpu_to_le16(byte_count);
 
 	rc = SendReceive(xid, tcon->ses, (struct smb_hdr *) pSMB,
 			 (struct smb_hdr *) pSMBr, &bytes_returned, 0);
